### DISCUSSION TITLE
Should the screenshot tool be deprecated?

### DISCUSSION THREAD

**Discussion by LWinterberg** [audacity/audacity#2336] [Development]
It is unclear to me whether the tool serves any practical purpose anymore - modern OSes typically have quite powerful screenshot tools built-in (Snipping tool since Windows Vista, [this](https://support.apple.com/guide/mac-help/take-a-screenshot-or-screen-recording-mh26782/mac) since at least High Sierra, Gnome/KDE/Xfce with [their respective tools](https://wiki.debian.org/ScreenShots)). There also are loads of advanced third party tools like Greenshot, ShareX, Snagit, etc. available, as well as some which are able to be used from a command line for scripting purposes (eg ImageMagick, Shutter or Scrot). 

My proposal would be

1. to mark the screenshot tool as deprecated, 
2. mark all bugs in #2335 as [Known Issue](../wiki/Known-Issues) 
3. keep the tool around until it actually causes trouble to someone (eg Paul's library extractions or the eventual replatforming)
4. when it eventually does cause trouble, remove it altogether.

Thoughts?

**Comment by SteveDaulton** [comment]
I've found the screenshot tool very useful in supporting users and documentation. It is particularly useful that it provides *consistent* images. Inconsistencies in screenshots can make documentation look shoddy, and if you just want, for example, a screenshot of a track, there's no quicker way than the screenshot tool.

It's also easier to direct a user to take a screenshot using Audacity's screenshot tool than having to remember how to take screenshots on multiple platforms.

It could be worth extending the screenshot tool to handle video if you are wanting video documentation.

**Reply by LWinterberg** [reply]
> I've found the screenshot tool very useful in supporting users and documentation. It is particularly useful that it provides consistent images. Inconsistencies in screenshots can make documentation look shoddy, and if you just want, for example, a screenshot of a track, there's no quicker way than the screenshot tool.

If it worked properly, yes. Due to #1608 and #1611, it is absolutely impossible for me to use the screenshot tool currently. It either captures the wrong region or nothing at all. 

> It could be worth extending the screenshot tool to handle video if you are wanting video documentation.

Video, GIF, drawing arrows, adding text labels, censoring sensitive/unnecessary parts, drawing boxes around important things - the list of things which are useful for documentation is long. It also is completely out-of-scope for an audio editor, so I don't think there should be many resources expended in this direction, especially considering that existing tools like ShareX do all of the above - including being able to make precise, one-click captures of many areas in Audacity, eg the transport toolbar:
![image](https://user-images.githubusercontent.com/87814144/147219528-22ec6fad-3f80-4ead-be05-3a03a7bbb151.png)

**Reply by SteveDaulton** [reply]
> If it worked properly, yes. Due to #1608 and #1611, it is absolutely impossible for me to use the screenshot tool currently. It either captures the wrong region or nothing at all.

It works fine for me when capturing a single track or the track panel, which are the only things that I routinely use it for. (I only use one monitor display).

> It also is completely out-of-scope for an audio editor

I don't think that it's out of scope for an audio editor to be able to take a screenshot of a track. Before the screenshot tool we only had the "File menu > Print" command, which has not worked well for as long as I've known it. We used to get complaints about poor results from the Print command, until we got the screenshot tool (a much simpler screenshot tool at that time). Complaints about the Print command ended when we got the screenshot tool, but there was clearly a demand for a quick and convenient way to capture waveforms and track spectrums as images.

The screenshot tool was greatly extended for the purpose of automating screenshots for the manual. As the manual is being abandoned I don't think there's a need for such a complex tool.

I agree that video capture is beyond the scope of an audio editor, but if you want users to produce video  documentation, then it may be helpful to provide users with the tools to do so. Video tutorials with Audacity expanded to fill a 4K screen are not very helpful for users with smaller displays, whereas a video capture tool could ensure that the Windows is scaled appropriately. Maybe worth considering, but just a thought - some features in Audacity are for the benefit of development / support / documentation, rather than directly for users.


I think that Paul's suggestion of extracting the screenshot tool into an optional module makes sense. Preferably the screenshot tool would be simplified so that it still fulfilled the needs of users, but without the *extra* features that were previously used for documentation.

**Comment by Paul-Licameli** [comment]
You should know that my monster branch of still unmerged refactoring work, called misc-libraries, extracts the screenshot tool (as it extracts so many other things) into an optional plug-in module.

All of the merged work, so far, has NOT yet included any extractions of high-level modules, which depend on the Audacity executable.  It has only been extraction of low-level libraries, which Audacity depends on.

**Comment by petersampsonaudacity** [comment]
I would vote for deprecating it - in fact I would vote for removing it entirely.

In all the images I captured for the manual over the years _(and there were very many)_ I never used the screenshot tool.  Rather I used the Windows Print Screen button and MS Paint to perform any trims or edits.

Mac too has perfectly good screen capture tools - albeit accessed by shortcuts and not a simple "discoverable" button on the keyboard.

If the screen shot tool were to be retained it would a lot of improvement, particularly in the GUI see #2306 - and for my money that would represent a poor ROI.

**Comment by Paul-Licameli** [comment]
I'm not invested much in the screenshot, but is it a point in favor of retaining it, that it is macro programmable?

Module separation in future might make it inexpensive to keep the thing around.  We might figure a way to deploy this and other little used features only optionally at the user's request, or at least to to load them by default at startup time.

**Comment by LWinterberg** [comment]
*@Penikov via #2335*
> I don't think we should rid of the Screenshot tool. There is a tool called UI Vision. It could be integrated into Jenkins CI.
We can set up Jenkins as a separate CI for automated tests. UI vision can capture the whole screen, but it works slow comprising a small image of the part of UI.
>Better if we just point UI Vision to a folder with a set of reference images. The screenshot tool provides that. Apart from that, UI Vision works more precisely with small pictures than with full-screen images.

>Yes, honestly, I do not use that tool in my daily testing. But if we talk about the coverage, developing Continuous Testing processes - that tool will be very useful.

**Comment by JamesCrook** [comment]
There are two screenshot dialogs.

There is the bloated dialog with many buttons.
There is also this short-format dialog:

![image](https://user-images.githubusercontent.com/1621525/147395159-03e0fb8c-0938-4d32-9f78-8a32662d897d.png)

The short-format tool is in the extra menu.  It is connected in to scripting.  The [demo scripts](https://github.com/audacity/audacity/tree/master/scripts/piped-work) walk through a lot of Audacity gathering screenshots.  I'm with Peter and LWinterberg that the large dialog should go from 'Tools'.  It is just bloat for users.   I'd think the short-format dialog should be maintained, unless you are already getting rid of scripting and the sample scripts. Then it can go too.

**Reply by petersampsonaudacity** [reply]
> There is also this short-format dialog:

So maybe the "bloated dialog with many buttons" in the Tools menu could just be replaced with this nice concise, clean-looking, format.


That would leave it being:
a) more "discoverable" (than hidden under scripting in the, off by default, Extra menu),
b) in the usual place for existing long-time users of the tool.

**Reply by JamesCrook** [reply]
I think it's better it just stays in the esoterica exotica extras menu.  Its main use is via scripting, and in scripting it makes some sense having a built in tool that is aware of the visual elements.  As a principle I think all the scriptable features should be available via menus too, so that's enough justification to keep the short-form dialog.  Remember that users can still screenshot using print screen or greenshot or other tools designed for screenshots.  It's only users doing a lot of screenshotting, and possibly automated, who benefit from the special screenshot feature.

About making it more elaborate and adding default path and browse button (copied from a config dialog) - I'm reluctant to see that.  Scripting is best when it exposes things raw, and the person writing the script uses features of python (say) to elaborate.  Go the cooked route, and you end up adding features like auto-naming and format strings for dates and times in the file name, features that make sense for an export multiple dialog (for audio) but not for a screenshot, which is a side feature of Audacity.

**Comment by petersampsonaudacity** [comment]
The downsides of the short form in the Extra menu are

1) There is no sensible default provided for the folder location for the screenshot.

2) Which means that you can get this not very helpful error message if you fail to enter a suitable location there:
![image](https://user-images.githubusercontent.com/50205984/147412512-74aded6e-3083-4f90-9d90-55e40eddd435.png)
This doesn't really properly tell the user what the problem is.

3) There is no Browse button to select a folder location _(to make life easier for the user)_.
